### DISCUSSION TITLE
Filter search results based on reference to other collection

### DISCUSSION THREAD

**Discussion by divvan** [ravendb/ravendb#16040] [Q&A] [answered]
I'd like to do a full text search of documents in one collection, and only return results that contain a particular field value in another collection.

I'm working on an application for factories that tracks problems with equipment. Say it has two classes with these string fields:

  * Equipment (Id, Name, Type)
  * Problem (Id, EquipmentId, Description)

The Equipment's Type field identifies its manufacturer and type. For example, a factory may use two different types of conveyor to move stuff around - "Intralox conveyor" and "Interroll conveyor". A large factory may have dozens or even hundreds of conveyor sections of each type. And despite the similarity in their names, their mechanisms are totally different, and so they can have different problems.

Now, lets say there's an Intralox conveyor called SortingConveyor_123, which is running noisily. The maintenance guy who has to fix this problem wants to know if this conveyor, or any other Intralox conveyor, has had the same problem before. 

So in the application, he searches for problems with the words "noisy Intralox conveyor". He may get no results because no other problem has the word "Intralox" in its Description, because people usually refer to equipment by name and not by type. Or he may get lots of results for noisy Interroll conveyors as well, but these are irrelevant because their mechanism is different.

Instead, I want the user to click on SortingConveyor_123 in the application and search for "noisy". Then it looks up that entity and returns all Problems related to equipment of type "Intralox conveyor" that have the word "noisy" in their Description.

But to achieve this, it can't just search in the Problem collection, because its documents only have the EquipmentId. Instead, it has to lookup each Problem's matching document in the Equipment collection to see if its Type is "Intralox conveyor". So, I basically need a Where clause in the search that joins to another collection.

How can I achieve this with RavenDB?

**Comment by ayende** [comment, marked as answer]
You can do that using `LoadDocument`, where when you index the `Problem` document, you can reach out to the related `Equipment` document.
See: 
https://ravendb.net/docs/article-page/5.4/csharp/indexes/indexing-related-documents
https://ayende.com/blog/179939/ravendb-4-0-unsung-heroes-indexing-related-data

**Comment by divvan** [comment]
Thanks, Ayende, that's a great feature!

I had thought of denormalizing Problem to include an EquipmentType field, but that would require patching if the related Type field was changed. And some other databases use materialized views, but they require more storage and they usually aren't updated automatically.

**Comment by divvan** [comment]
I don't think the LoadDocument method will work any more due to recent changes to the data model that have added more relationships:

![Problem](https://github.com/user-attachments/assets/202d9b02-d10d-430c-b330-ff5b2bb8ae85)

The Problem entity no longer references the Equipment entities directly,, but there's a new Status entity that records the status of the problem for each Equipment entity. And users can add many Comments about the problem, instead of the single Description field I mentioned in my original post.

So, this is what I'm trying to achieve with this new model: 

Say a factory has one or more Intralox conveyors that have become noisy. Those conveyors are represented by Equipment entities whose Type field is set to "Intralox conveyor". I want to do a full text search of the Comments from past Problems to see if Intralox conveyors have been noisy before. So, I want to filter out Comments for other Equipment Types (such as Interroll conveyors), even if they have the word "noisy".

So, can I still filter the search results based on the Equipment Type field with all these extra relationships?

To reduce the number of relationships, I've thought of embedding the Status and Comment in the Problem entity. Although I could do that with the Status, I can't do it with the Comment. Also, the model is a bit more complicated than this as there's another entity that references Equipment (which I haven't shown here) that is also used in the search.

**Reply by ayende** [reply]
I assume that each one of those is a _separate_ document?

Is there a reason that there isn't also a link between `Comment` and the relevant `Equipment` (or multiple such) ?

**Reply by divvan** [reply]
Thanks for you reply, Oren.

Yes, most of these entities need to be separate documents.

Due to other constraints, the Comment document can't have links to all the Equipment documents the Problem  references. But I could create a new entity that contains those links plus the text from the Comment. Then the full text search would use this entity instead of the Comment.

But the main downside is that there could be a lot of duplication of those links. For example, if the Problem references 3 Equipment entities and it has 20 Comments, there would be 60 links instead of the original 3.

But if there's no other option...

**Reply by ayende** [reply]
Do it the other way around.
Index the `Problem` document, which will have a list of `Comments` - then you can pull of that together,no?

**Comment by divvan** [comment]
Is it possible to do an AND operation that combines search results from two separate related documents?

I want to return Comments whose Text field contains the specified terms (such as "noisy"), when the related Equipment document's Type field contains different terms (such as "Intralox conveyor"). 

So, if I start with the Status document that's in the middle of the diagram above, can I do two separate searches on its related Problem/Comment and Equipment documents, and somehow combine their results with an AND operator?

**Reply by ayende** [reply]
So, you can use `LoadDocument` for that, see https://ravendb.net/docs/article-page/7.0/csharp/indexes/indexing-related-documents

**Comment by divvan** [comment]
Sorry, I'm not explaining this very well.

I know I can use LoadDocument to create an index based on the contents of a related document. But in all the search examples I've seen in the documentation, a search is only based on the contents of one collection. But I want to do something like an Intersect operation between the indexes on two separate collections that share the same Status document.

For example, say I use LoadDocument to create these indexes:

	Status_byProblemCommentText
	Status_byEquipmentType

Then I want to do something like this:

	select Comment from Status_byProblemCommentText
	where Comment.Text = "noisy"
	intersect Status_byEquipmentType 
	where Equipment.Type = "Intralox conveyor"
	
So, it will only return a Comment if these conditions are true for both the Comment and Equipment documents referenced by the Status document.

**Reply by georgiosd** [reply]
You need a multi-map index. https://ravendb.net/docs/article-page/7.0/csharp/indexes/multi-map-indexes

There will need be some information that links the records, like you would do with a JOIN in SQL.

Otherwise you can issue two queries lazily and join the results in memory (not as optimal).
